vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/treegrid/TreeGrid.java

```diff
@@ -155,56 +157,37 @@ public TreeGridUpdateQueue startUpdate(int sizeChange) {
                 viewportRemaining = EAGER_FETCH_VIEWPORT_SIZE_ESTIMATE;
+                queuedParents.clear();
                 previousRequest = VaadinRequest.getCurrent();
             }
 
-            queue.setArrayUpdateListener((start, length, parentKey) -> {
-                if (viewportRemaining > 0) {
+            queue.setArrayUpdateListener((items) -> {
+                // Prepend the items to the queue of potential parents.
+                queuedParents.addAll(0, items);
 
-                    T parentItem = parentKey == null ? null
-                            : getDataCommunicator().getKeyMapper()
-                                    .get(parentKey);
+                while (viewportRemaining > 0 && !queuedParents.isEmpty()) {
+                    viewportRemaining--;
+                    JsonObject parent = (JsonObject) queuedParents.remove(0);
+                    T parentItem = getDataCommunicator().getKeyMapper()
+                            .get(parent.getString("key"));
 
-                    viewportRemaining = recursiveSetParentRequestedRange(start,
-                            length, parentItem, viewportRemaining);
-                }
-            });
-
-            return queue;
-        }
-
-        private int recursiveSetParentRequestedRange(int start, int length,
-                T parentItem, int viewportRemaining) {
-
-            // Make a query for the item children
-            HierarchicalQuery<T, SerializablePredicate<T>> query = new HierarchicalQuery<>(
-                    start, getPageSize(),
-                    getDataCommunicator().getBackEndSorting(),
-                    getDataCommunicator().getInMemorySorting(), null,
-                    parentItem);
+                    if (isExpanded(parentItem)) {
+                        int childLength = Math.max(
+                                EAGER_FETCH_VIEWPORT_SIZE_ESTIMATE,
+                                getPageSize());
 
-            List<T> children = getDataProvider().fetchChildren(query)
-                    .collect(Collectors.toList());
+                        // There's still room left in the viewport and the item
+                        // is expanded. Set parent requested range for it.
+                        getDataCommunicator().setParentRequestedRange(0,
+                                childLength, parentItem);
 
-            for (T child : children) {
-                viewportRemaining -= 1;
-
-                if (viewportRemaining > 0 && isExpanded(child)) {
-                    int childLength = Math.max(
-                            EAGER_FETCH_VIEWPORT_SIZE_ESTIMATE, getPageSize());
-
-                    // There's still room left in the viewport and the child is
-                    // expanded. Set parent requested range for the child.
-                    getDataCommunicator().setParentRequestedRange(0,
-                            childLength, child);
+                        break;
```

Comment:
Got stuck a bit on this part until I realized that requesting children triggers another while loop, and that would then just keep processing the parent siblings that might be left in the queue. Maybe we could add some explanation here.

There is one scenario where this fails, which is if the parent is expanded, but has no children. I'm not 100% certain how the data provider contract works, but I can reproduce it by overriding the following methods in the custom `LazyHierarchicalDataProvider` in `TreeGridPreloadPage`, and then opening http://localhost:8080/vaadin-grid/treegrid-preload?expandedRootIndexes=1:
```java
                    @Override
                    public boolean hasChildren(HierarchicalTestBean item) {
                        if(item.getId().equals("/0/1/1/0")) {
                            return true;
                        }
                        return super.hasChildren(item);
                    }

                    @Override
                    public int getChildCount(HierarchicalQuery<HierarchicalTestBean, Void> query) {
                        if(query.getParentOptional().isPresent() && query.getParent().getId().equals("/0/1/1/0")) {
                            return 0;
                        }
                        return super.getChildCount(query);
                    }

                    @Override
                    public Stream<HierarchicalTestBean> fetchChildren(HierarchicalQuery<HierarchicalTestBean, Void> query) {
                        if(query.getParentOptional().isPresent() && query.getParent().getId().equals("/0/1/1/0")) {
                            return Stream.of();
                        }
                        return super.fetchChildren(query);
                    }
```

That leads to multiple round-trips then:

![Bildschirmfoto 2022-03-04 um 13 33 22](https://user-images.githubusercontent.com/357820/156764349-6f6555bb-45a1-4f3f-a45e-2717f67262e5.png)

That wouldn't be a huge problem, but maybe we can avoid that somehow.

Reply:
> Maybe we could add some explanation here.

Added the documentation

> There is one scenario where this fails, which is if the parent is expanded, but has no children

I guess this would be quite an edge case and multiple roundtrips wouldn't be too serious of an issue if this happens to occur. Nonetheless, added a check for `hasChildren`.

Reply:
> Nonetheless, added a check for hasChildren.

After some more thinking, decided to revert this. Depending on how `hasChildren` is implemented in the `DataProvider`, an additional server roundtrip on a very rare occasion is probably a lesser evil than running `hasChildren` for each child item every time.

Reply:
Seems reasonable, let's keep it like that.